File: Learning Codes/Linked List/Mine/Python/delete.py
class Node:
    def linked_list(self, val):
        self.data = val
        self.next = None

def insert_at_tail(head, val):
    n = Node()
    n.linked_list(val)
    if head is None:
        head = n
        return head  # Return the updated head
    tmp = head
    while tmp.next is not None:
        tmp = tmp.next
    tmp.next = n
    return head  # Return the updated head

def delete_at_position(head, position):
    if head is None:
        print("List is empty. Cannot delete.")
        return None  # Return None if list is empty

    tmp = head
    if position == 1:
        head = head.next
        del tmp
        return head

    i = 1
    while i < position - 1 and tmp is not None:
        tmp = tmp.next
        i += 1

    if tmp is None or tmp.next is None:
        print("Position out of range. Cannot delete.")
        return head

    to_delete = tmp.next
    tmp.next = to_delete.next

    if tmp.next is None:
        del to_delete
        return head

    del to_delete
    return head  

def display(head):
    tmp = head
    while tmp is not None:
        print(tmp.data, "-->", end=" ")
        tmp = tmp.next
    print("NULL")

head = None
head = insert_at_tail(head, 1)
head = insert_at_tail(head, 10)
head = insert_at_tail(head, 100)
head = insert_at_tail(head, 1000)
display(head)

head = delete_at_position(head, 4)
display(head)

File: Learning Codes/Linked List/Mine/C++/delete.cpp
#include <bits/stdc++.h>
using namespace std;

class node
{
public:
    int data;
    node *next;

    node(int val)
    {
        data = val;
        next = NULL;
    }
};

void insert_at_tail(node *&head, int val)
{
    node *n = new node(val);
    if (head == NULL)
    {
        head = n;
        return;
    }
    node *tmp = head;
    while (tmp->next != NULL)
    {
        tmp = tmp->next;
    }
    tmp->next = n;
}

void delete_at_position(node *&head, int position)
{
    if (head == NULL)
    {
        cout << "List is empty. Cannot delete." << endl;
        return;
    }

    node *tmp = head;
    if (position == 1)
    {
        head = head->next;
        delete tmp;
        return;
    }

    int i = 1;
    while (i < position - 1 && tmp != NULL)
    {
        tmp = tmp->next;
        i++;
    }

    if (tmp == NULL || tmp->next == NULL)
    {
        cout << "Position out of range. Cannot delete." << endl;
        return;
    }

    node *toDelete = tmp->next;
    tmp->next = toDelete->next;

    if (tmp->next == NULL)
    {
        delete toDelete;
        return;
    }
    delete toDelete;
}

void display(node *head)
{
    node *tmp = head;
    while (tmp != NULL)
    {
        cout << tmp->data << " --> ";
        tmp = tmp->next;
    }
    cout << "NULL" << endl;
}

int main()
{
    node *head = NULL;
    insert_at_tail(head, 1);
    insert_at_tail(head, 10);
    insert_at_tail(head, 100);
    insert_at_tail(head, 1000);
    display(head);

    delete_at_position(head, 4);
    display(head);

    return 0;
}
